main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -85,8 +89,21 @@ object Inliner {
       Expression.ApplyClo(e, as, tpe, loc)
 
     case Expression.ApplyDef(sym, args, tpe, loc) =>
-      val as = args.map(visitExp(_, subst0))
-      Expression.ApplyDef(sym, as, tpe, loc)
+      val as1 = args.map(visitExp(_, subst0))
+      val def0 = root.defs.apply(sym) // f(x: Int)
+      // If `sym` is a single non-self call and is trivial, then inline `sym`.
+      if (def0.occurDef.isTrivialNonSelfCall) {
+        // Map parameters of the function `def0` to its arguments in the substitution map `subst1`.
+        val subst1 = subst0 ++ def0.fparams.map(_.sym).zip(as1).toMap
```

Comment:
What you are doing here is really multiple things:

- You are trying to ensure that you don't mistakenly insert a tail call in a non-tail position.
- You are binding the formal arguments to their actual parameters.
- You are applying a form of inlining by doing in the environment.

I would propose to break this down into a few helper functions:

- You can write a function that takes an expression and converts tail calls to non-tail calls. 
- Instead of binding formals to actuals via subst, you could bind them via new let-bindings:

```
def f(x: Int32): Int32 = x

def g(): Int32 = f(123)
```

becomes

```
def g(): Int32 = 
  let tmp1 = 123;
  tmp1 // by taking the body of f and replacing f[x -> tmp1].
```

Why do you think I propose this? On the surface it would seem like a worse solution.